File: android/app/src/main/java/com/noop/ui/AlertInbox.kt
package com.noop.ui

import android.content.Context
import java.time.Instant
import java.time.ZoneId

/**
 * Records alerts that really happened in the app's bell. It deliberately does not represent scheduled
 * reminders: those can be delivered while the process is absent and must not be claimed as events early.
 */
object AlertInbox {
    enum class Kind(
        val key: String,
        val lifetimeMs: Long,
        val deepLink: String? = null,
    ) {
        BATTERY_LOW("battery-low", 3L * 24L * 60L * 60L * 1000L, "devices"),
        BATTERY_RUNTIME("battery-runtime", 3L * 24L * 60L * 60L * 1000L, "devices"),
        BATTERY_FULL("battery-full", 24L * 60L * 60L * 1000L, "devices"),
        BATTERY_CRITICAL("battery-critical", 3L * 24L * 60L * 60L * 1000L, "devices"),
        BATTERY_BEDTIME("battery-bedtime", 24L * 60L * 60L * 1000L, "devices"),
        ILLNESS("illness", 2L * 24L * 60L * 60L * 1000L),
        INACTIVITY("inactivity", 24L * 60L * 60L * 1000L),
        SMART_ALARM("smart-alarm", 24L * 60L * 60L * 1000L),
    }

    fun post(context: Context, kind: Kind, title: String, message: String, now: Long = System.currentTimeMillis()) {
        val day = Instant.ofEpochMilli(now).atZone(ZoneId.systemDefault()).toLocalDate()
        val key = "${kind.key}:$day"
        UpdateStore.from(context).postOrRefreshAlert(
            alertKey = key,
            title = title,
            message = message,
            deepLink = kind.deepLink,
            expiresAt = now + kind.lifetimeMs,
            now = now,
        )
    }
}

File: Strand/System/AlertInbox.swift
import Foundation

/// Records recent, user-facing local alerts in the bell without coupling the alert engines to SwiftUI.
/// This is intentionally separate from scheduled reminders: only an event that actually happened calls
/// this bridge. A denied system-notification permission therefore hides the banner, never the local log.
enum AlertInbox {
    enum Kind: String {
        case batteryLow = "battery-low"
        case batteryRuntime = "battery-runtime"
        case batteryFull = "battery-full"
        case batteryCritical = "battery-critical"
        case batteryBedtime = "battery-bedtime"
        case illness = "illness"
        case inactivity = "inactivity"
        case smartAlarm = "smart-alarm"

        var lifetime: TimeInterval {
            switch self {
            case .batteryLow, .batteryRuntime, .batteryCritical: return 3 * 24 * 60 * 60
            case .batteryFull, .batteryBedtime:                  return 24 * 60 * 60
            case .illness:                     return 2 * 24 * 60 * 60
            case .inactivity, .smartAlarm:     return 24 * 60 * 60
            }
        }

        var deepLink: String? {
            switch self {
            case .batteryLow, .batteryRuntime, .batteryFull, .batteryCritical, .batteryBedtime:
                return NavRouter.Destination.devices.rawValue
            case .illness, .inactivity, .smartAlarm: return nil
            }
        }
    }

    /// Safe from the BLE and notification callbacks, which do not all run on the main actor.
    static func post(_ kind: Kind, title: String, message: String, now: Date = Date()) {
        let key = "\(kind.rawValue):\(dayKey(now))"
        let expiresAt = now.addingTimeInterval(kind.lifetime)
        Task { @MainActor in
            UpdateStore.shared.postOrRefreshAlert(
                key: key, title: title, message: message,
                deepLink: kind.deepLink, expiresAt: expiresAt, now: now
            )
        }
    }

    private static func dayKey(_ date: Date) -> String {
        let f = DateFormatter()
        f.calendar = .current
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyy-MM-dd"
        return f.string(from: date)
    }
}
